### PR TITLE
Add register dialog to separate login/register

### LINKED ISSUE #8138: Separate register and login
Change logins to work like so:

```
| server list            | | address       | | port    |
|                        | | server description        |
|                        | |                           |
|                        | |                           |
|                        | |                           |
|                        | |                           |
|                        | |                           |
|                        | | username    | | password  |
|                        | | Login      | |  Register  |
```

1. To login, enter your username and password on the mainmenu and click login
2. To register, click register to open the register dialog then enter your username and password twice on the dialog
3. When clicking login, if the account doesn't exist then ~it redirects to the register dialog~ it is registered anyway
4. When attempting to register an existing username, it should give an error

This should make the per-server account more obvious

EDIT by pararmat: For #258 and #7237

### PR DESCRIPTION
**Pull request by rubenwardy**
Fixes #8138

New users find Minetest's account system confusing. This change moves username/password to a new dialog, with login and register buttons added to the Join Game tab.

The old registration confirmation dialog is removed in favour of the new dialog. Instead of disabling register confirmation, users can completely disable split login/register using `enable_split_login_register = false`

image
image
image
image

## To do

- [x] Move delete favourite button to a better location
- [x] Error messages on register from login, and login from register
- [x] What to do when connecting to localhost?

## How to test

* Register:
    * Clear minetest.conf. Click register, confirm name field is empty and focused.
    * Fill in information, connect. Confirm server is marked as favourite.
    * Attempt to register on another server, confirm that name is remembered and password is focused
    * Test behaviour when passwords do not match
* Login:
    * Test correct log-in
    * Test attempt to register
    * Test incorrect password
* Test `enable_split_login_register = false` behaves as master, but without the register confirmation dialog

### CODE REVIEW

**Comment by erlehmann** [comment]
> One caveat is that it's possible to register from login, and login from register. I believe this will require network protocol changes in order to detect/separate login from register, I'll need @sfan5's guidance on that. Once that is done, it would be nice to show better errors on the dialog - for example, saying "that username is already taken" rather than access denied on the register from.

I believe that looking at the `TOCLIENT_HELLO` packet sent by the server makes this already possible without changes to the network protocol. If the registration mechanism is `AUTH_MECHANISM_FIRST_SRP`, the username is available – otherwise, it is not. However, i have not thought about this too long; I spent maybe 10 minutes on reading the code before I made this suggestion.

(Fortunately, this is really easy to test.)

**Comment by sfan5** [comment]
Concept: Sounds good, doesn't add an extra step to log in as existing user.

UI:
* Should be no gap between Name + Password
* Blue text vaguely suggests that it is clickable but why is Register not a button?
* It becomes impossible to unfavorite a server :skull: 

Text:
* "Name not registered" maybe add something like "If you want to join this server for the first time click on 'Register'"
* "Join example.com" -> "Joining example.com:"

Code: didn't look yet

**Comment by rubenwardy** [comment]
> Blue text vaguely suggests that it is clickable but why is Register not a button?

This was something I was experimenting with, but I think it looks out of place. The idea is to indicate that it's a navigation rather than an action, as this is usually the difference between links and buttons. This isn't done anywhere else though, and looks out of place, so I'll revert it to a button

> It becomes impossible to unfavorite a server skull

See the favourites star next to "Server description"

**Comment by sfan5** [comment]
> See the favourites star next to "Server description"

It needs to actually look like a button for that to work.
Also here's an icon suggestion:
grafik

**Comment by rubenwardy** [comment]
Done

image

File: builtin/mainmenu/tab_online.lua [hunk 1]
```diff
@@ -87,27 +87,34 @@ local function get_formspec(tabview, name, tabdata)
...
+
+	if core.settings:get_bool("enable_split_login_register") then
+		retval = retval .. "button[0.25,6;2.5,0.75;btn_mp_register;" .. fgettext("Register") .. "]"
+	end
 
 	if tabdata.selected then
 		if gamedata.fav then
-			retval = retval .. "button[0.25,6;2.5,0.75;btn_delete_favorite;" ..
-				fgettext("Del. Favorite") .. "]"
+			retval = retval .. "tooltip[btn_delete_favorite;" .. fgettext("Delete favorite") .. "]"
```
Comment (sfan5) [inline builtin/mainmenu/tab_online.lua:111]:
```suggestion
			retval = retval .. "tooltip[btn_delete_favorite;" .. fgettext("Remove favorite") .. "]"
```

File: builtin/mainmenu/dlg_register.lua [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+--GNU Lesser General Public License for more details.
+--
+--You should have received a copy of the GNU Lesser General Public License along
+--with this program; if not, write to the Free Software Foundation, Inc.,
+--51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+--------------------------------------------------------------------------------
+
+local function register_formspec(dialogdata)
+	local title = fgettext("Joining $1", dialogdata.address)
```
Comment (sfan5) [inline builtin/mainmenu/dlg_register.lua:21]: maybe use the server name here if available

File: builtin/mainmenu/dlg_register.lua [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		"field[0.375,1.575;7.25,0.8;name;", core.formspec_escape(fgettext("Name")), ";",
+				core.formspec_escape(dialogdata.name), "]",
+		"pwdfield[0.375,2.875;7.25,0.8;password;", core.formspec_escape(fgettext("Password")), "]",
+		"pwdfield[0.375,4.175;7.25,0.8;password_2;", core.formspec_escape(fgettext("Confirm Password")), "]"
+	}
+
+	if dialogdata.error then
+		table.insert_all(retval, {
+			"style_type[label;textcolor=", mt_color_orange, "]",
+			"label[0.375,", tostring(buttons_y - 0.6), ";", core.formspec_escape(dialogdata.error), "]",
```
Comment (sfan5) [inline builtin/mainmenu/dlg_register.lua:41]: I think a `box[]` would be nice around the error message.

File: builtin/mainmenu/dlg_register.lua [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		local server = this.data.server
+		if server then
+			serverlistmgr.add_favorite(server)
+			gamedata.servername        = server.name
+			gamedata.serverdescription = server.description
+
+			if not is_server_protocol_compat_or_error(
+						server.proto_min, server.proto_max) then
+				return true
+			end
```
Comment (sfan5) [inline builtin/mainmenu/dlg_register.lua:88]: This should be checked before the register dialog is shown

**Review by sfan5 — COMMENTED** [review]
Works other than that.

File: src/script/lua_api/l_mainmenu.cpp [hunk 5]
```diff
@@ -139,6 +139,14 @@ int ModApiMainMenu::l_start(lua_State *L)
...
+		const auto val = getTextData(L, "allow_login_or_register");
+		if (val == "login")
+			data->allow_login_or_register = ELoginRegister::Login;
+		else if (val == "register")
+			data->allow_login_or_register = ELoginRegister::Register;
+		else
+			data->allow_login_or_register = ELoginRegister::Any;
 	}
 	data->serverdescription = getTextData(L,"serverdescription");
 	data->servername        = getTextData(L,"servername");
```
Comment (sfan5) [inline src/script/lua_api/l_mainmenu.cpp:152]: also the PNG isn't optimized, again ;)

File: builtin/mainmenu/dlg_register.lua (hunk 3 above)
Comment (rubenwardy) [reply builtin/mainmenu/dlg_register.lua:41]: Which color?  A darker red with white text?

**Comment by rubenwardy** [comment]
image

**Comment by sfan5** [comment]
Maybe not that bright but sure.

**Comment by Zughy** [comment]
@sfan5 if it were brighter, people would struggle reading the (white) text. If it were gray, it'd resemble an input field. I think this choice is a nice compromise between readability and UX (also considering the whole meh default aspect of formspecs)

**Review by SmallJoker — APPROVED** [review]
--> Icon suggestion: server_favorite_delete greytone + 1 simple stroke to highlight the deletion aspect.

`enable_split_login_register = true`:

:ballot_box_with_check: Register: empty password + "disallow_empty_password=true" -> "Protocol error"
:ballot_box_with_check: Register: normal password
:ballot_box_with_check: Login: unknown account -> "Protocol error"
:ballot_box_with_check: Login: correct password
:ballot_box_with_check: Login: wrong password -> "Protocol error"

`enable_split_login_register = false`:

:ballot_box_with_check: New account
:ballot_box_with_check: Login: correct password
:ballot_box_with_check: Login: wrong password -> "Protocol error"

The PR works, ~~but the recent error message changes are definitely not good.~~ PR with a fix is pending (#12378).

**Comment by sfan5** [comment]
As discussed on IRC the license info of the icon needs to be noted somewhere.

**Review by sfan5 — APPROVED** [review]

**Pull request merged**